controllers/kubeadmconfig_controller.go

```diff
@@ -151,12 +151,11 @@ func (r *KubeadmConfigReconciler) Reconcile(req ctrl.Request) (_ ctrl.Result, re
 	if err != nil {
 		return ctrl.Result{}, err
 	}
-	// Always attempt to Patch the KubeadmConfig object and status after each reconciliation.
+	// Attempt to Patch the KubeadmConfig object and status after each reconciliation if no error occurs.
 	defer func() {
-		if err := patchHelper.Patch(ctx, config); err != nil {
-			log.Error(err, "failed to patch config")
-			if rerr == nil {
-				rerr = err
+		if rerr == nil {
```

Comment:
This seems like a departure from other cluster-api related use cases, where we would potentially want to ensure that we persist some type of intermediate or partial state back to the resource. That said, the behaviors of this controller vs those other cases it is probably generally ok.

Are we attempting to set ErrorMessage and ErrorReason anywhere where this change could potentially prevent those changes from persisting?

Reply:
We do not currently use ErrorMessage or ErrorReason in this provider